libs/vectorise/include/vectorise/uint/uint.hpp

```diff
@@ -1081,39 +1138,69 @@ UInt<S>::operator std::string() const
 }
 
 template <uint16_t S>
-inline std::ostream &operator<<(std::ostream &s, UInt<S> const &x)
+template <typename T>
+constexpr meta::EnableIf<std::is_same<meta::Decay<T>, byte_array::ByteArray>::value, T> UInt<S>::As(
+    bool as_little_endian, bool include_leading_zeroes) const
+{
+  auto const size_{include_leading_zeroes ? size() : TrimmedSize()};
+
+  if (0 == size_)
+  {
+    return T{};
+  }
+
+  if (as_little_endian)
+  {
+    return T{pointer(), size_};
+  }
+
+  T array;
+  array.Resize(size_);
+  for (std::size_t i{0}, rev_i{size_ - 1}; i < size_; ++i, --rev_i)
+  {
+    array[rev_i] = base()[i];
+  }
+
+  return array;
+}
+
+template <uint16_t S>
+std::ostream &operator<<(std::ostream &s, UInt<S> const &x)
 {
   s << std::string(x);
   return s;
 }
 
-inline double ToDouble(UInt<256> const &x)
+template <typename T>
+constexpr meta::EnableIf<std::is_same<meta::Decay<T>, UInt<256>>::value, double> ToDouble(T &&x)
 {
-  if (x.TrimmedSize() == 1)
+  auto const msw_size{x.TrimmedWideSize()};
+  if (0 == msw_size)
   {
-    return static_cast<double>(x.ElementAt(0));
+    return 0;
   }
-  const size_t ELEMENTS_PER_WIDE = x.WIDE_ELEMENT_SIZE / x.ELEMENT_SIZE;
-  uint64_t     magnitude         = x.ELEMENTS * x.ELEMENT_SIZE - x.msb() - 1;
 
-  const size_t most_significant_byte_idx = magnitude / x.ELEMENT_SIZE;
-  const size_t least_eligible_byte_idx   = most_significant_byte_idx - (ELEMENTS_PER_WIDE - 1);
+  const auto     msw_index{msw_size - 1};
+  const uint64_t most_significant_word = x.ElementAt(msw_index);
+  auto           retval = pow(2, static_cast<double>(msw_index * x.WIDE_ELEMENT_SIZE)) *
+                static_cast<double>(most_significant_word);
 
-  uint64_t mantisse = 0;
-  for (size_t i = 0; i < ELEMENTS_PER_WIDE; ++i)
+  if (msw_index > 0 and most_significant_word < (1ull << 49))
   {
-    const uint64_t element  = x[i + least_eligible_byte_idx];
-    const uint64_t addendum = uint64_t(element) << (i * x.ELEMENT_SIZE);
-    mantisse += addendum;
+    auto const     next_idx{msw_index - 1};
+    uint64_t const word2 = x.ElementAt(next_idx);
+    auto           v2 =
+        pow(2, static_cast<double>(next_idx * x.WIDE_ELEMENT_SIZE)) * static_cast<double>(word2);
+    retval += v2;
   }
-  const auto exponent = static_cast<double>(least_eligible_byte_idx * x.ELEMENT_SIZE);
 
-  return static_cast<double>(mantisse) * pow(2, exponent);
+  return retval;
 }
 
-inline double Log(UInt<256> const &x)
+template <typename T>
+constexpr meta::EnableIf<std::is_same<meta::Decay<T>, UInt<256>>::value, double> Log(T &&x)
```

Comment:
I am not sure if we should have this log function on the UInt256. Since we are converting to double. Does it makes sense to simply include a count leading zeros function or similar? (and then people can use it as a psuedo log or similar)

Reply:
The method was there before.
Wanted to ditch it exactly due to reasons you mentioned (we have to double conversion), but decided to keep it just in case there is some Etch code which relies on it (e.g. some smart/synergetic contract tests, examples of code in the documentation, etc. ...).
I will be more than happy to ditch it.